packages/gatsby/src/redux/reducers/queries.ts

```diff
@@ -0,0 +1,233 @@
+      return state
+    }
+    case `DELETE_PAGE`: {
+      // Don't actually remove the page query from trackedQueries, just mark it as "deleted". Why?
+      //   We promote a technique of a consecutive deletePage/createPage calls in onCreatePage hook,
+      //   see https://www.gatsbyjs.com/docs/creating-and-modifying-pages/#pass-context-to-pages
+      //   If we remove a query and then re-add, it will be marked as dirty.
+      //   This is OK for cold cache but with warm cache we will re-run all of those queries (unnecessarily).
+      //   We will reconcile the state after createPages API call and actually delete those queries.
+      state.deletedQueries.add(action.payload.path)
+      return state
+    }
+    case `API_FINISHED`: {
+      if (action.payload.apiName !== `createPages`) {
+        return state
+      }
+      for (const queryId of state.deletedQueries) {
+        for (const component of state.trackedComponents.values()) {
+          component.pages.delete(queryId)
+        }
+        for (const nodeQueries of state.byNode.values()) {
+          nodeQueries.delete(queryId)
+        }
+        for (const connectionQueries of state.byConnection.values()) {
+          connectionQueries.delete(queryId)
+        }
+        state.trackedQueries.delete(queryId)
+      }
+      state.deletedQueries.clear()
+      return state
+    }
+    case `QUERY_EXTRACTED`: {
+      // Note: this action is called even in case of
+      // extraction error or missing query (with query === ``)
+      // TODO: use hash instead of a query text
+      const { componentPath, query } = action.payload
+      const component = registerComponent(state, componentPath)
+      if (component.query !== query) {
+        // Invalidate all pages associated with a component when query text changes
+        component.pages.forEach(queryId => {
+          const query = state.trackedQueries.get(queryId)
+          if (query) {
+            query.dirty = setFlag(query.dirty, FLAG_DIRTY_TEXT)
+          }
+        })
+        component.query = query
+      }
+      return state
+    }
+    case `QUERY_EXTRACTION_GRAPHQL_ERROR`:
+    case `QUERY_EXTRACTION_BABEL_ERROR`:
+    case `QUERY_EXTRACTION_BABEL_SUCCESS`: {
+      const { componentPath } = action.payload
+      const component = registerComponent(state, componentPath)
+      const set = action.type !== `QUERY_EXTRACTION_BABEL_SUCCESS`
+      component.errors = setFlag(component.errors, FLAG_ERROR_BABEL, set)
+      return state
+    }
+    case `REPLACE_STATIC_QUERY`: {
+      // Only called when static query text has changed, so no need to compare
+      // TODO: unify the behavior?
+      const query = registerQuery(state, action.payload.id)
+      query.dirty = setFlag(query.dirty, FLAG_DIRTY_TEXT)
+      state.deletedQueries.delete(action.payload.id)
+      return state
+    }
+    case `REMOVE_STATIC_QUERY`: {
+      state.deletedQueries.add(action.payload)
+      return state
+    }
```

Comment:
Do we flush the `deletedQueries` for static query only code paths? From what I can tell only flush happen on `createPages` api finished event - is this guaranteed to happen? If it's not guaranteed - could this cause problems?

I wonder if we need to trigger additional flush after query extraction? And also if it would need separate `deletedQueries` for page and static queries?

Reply:
We also flush `deletedQueries` in `CREATE_PAGE` action here:

https://github.com/gatsbyjs/gatsby/blob/cb1536ed49191638f019454fb9e3c225d776f3db/packages/gatsby/src/redux/reducers/queries.ts#L62-L70

So the only undesired effect is that it may keep some garbage in memory until the next `createPages` API call. Functionality-wise I don't see how it can cause issues.